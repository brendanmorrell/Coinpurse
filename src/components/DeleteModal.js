import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';

import { startRemoveExpense } from '../actions/expenses';
import { toggleModal, pushToDashboard } from '../actions/modal';

export class DeleteModal extends React.Component {
  onRemoveExpense = () => {
    const { modal } = this.props;
    const { id } = this.props;
    this.props.startRemoveExpenseProp(id);
    this.props.toggleModal(!modal);
    this.props.pushToDashboard(true);
  };
  doNotRemoveExpense = () => {
    const { modal } = this.props;
    this.props.toggleModal(!modal);
  };
  render() {
    return (
      <div>
        <Modal
          isOpen={this.props.modal}
          className="modal"
          onAfterOpen={() => ({})}
          onRequestClost={() => ({})}
          closeTimeoutMS={11100}
          contentLabel="Modal"
        >
          <h3 className="modal__title">Are you sure you want to delete this expense?</h3>
          <div className="modal-div">
            <button className="button" onClick={this.onRemoveExpense}>Yes</button>
            <button className="button" onClick={this.doNotRemoveExpense}>No</button>
          </div>
        </Modal>
      </div>
    );
  }
}


const mapDispatchToProps = (dispatch, props) => ({
  startRemoveExpenseProp: (id) => dispatch(startRemoveExpense(id)),
  toggleModal: (modal) => dispatch(toggleModal(modal)),
  pushToDashboard: (pushTD) => dispatch(pushToDashboard(pushTD)),
});
const mapStateToProps = (state, props) => ({ modal: state.modal.modal });
export default connect(mapStateToProps, mapDispatchToProps)(DeleteModal);
