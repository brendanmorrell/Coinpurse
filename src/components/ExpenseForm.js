import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';// this is an airbnb available module that gives a calendar. other versions also available

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      errorDescription: '',
      errorAmount: '',
      errorDate: '',
      dateMessage: '',
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description, errorDescription: '' }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      return this.setState(() => ({ amount, errorAmount: '' }));
    }
    return undefined;
  };
  onDateChange = (createdAt) => { // this function takes a moment input
    this.setState(() => ({ createdAt, errorDate: '' }));
    if (!moment(createdAt)._isValid) {
      return this.setState(() => ({ dateMessage: 'DD/MM/YYYY' }));
    }
    return this.setState(() => ({ dateMessage: '' }));
  };
  onFocusChange = ({ focused }) => { // destructured somehow. not sure where '.focused' is getting pulled from. maybe the focused property in the <SingleDatePicker component?
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount || !this.state.createdAt) {
      if (!this.state.description) {
        this.setState(() => ({ errorDescription: 'Description Required' }));
      }
      if (!this.state.amount) {
        this.setState(() => ({ errorAmount: 'Amount Required' }));
      }
      if (!this.state.createdAt) {
        this.setState(() => ({ errorDate: 'Dates must be formatted DD/MM/YYYY' }));
      }
      return;
    }
    this.props.onSubmit({
      description: this.state.description,
      amount: parseFloat(this.state.amount, 10) * 100, // turns the string into a real number (opposite is parseInt). second arg is the base 10, and the third is because we want cents
      createdAt: this.state.createdAt.valueOf(), // value of is a moment function that takes a moment object and returns it to a unix epock milliseconds value
      note: this.state.note,
    });
  }
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.errorDescription && <p className="form__error">{this.state.errorDescription}</p>}
        {this.state.errorAmount && <p className="form__error">{this.state.errorAmount}</p>}
        {this.state.errorDate && <p className="form__error">{this.state.errorDate}</p>}
        <input
          type="text"
          placeholder="Description"
          autoFocus
          className="text-input"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          className="text-input"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker// date ondatechange focused and onfocuschange are all required props of the component. there are additional optional ones you can add too
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          // the below props are optional
          numberOfMonths={1}
          isOutsideRange={() => false}
        />{this.state.dateMessage && <p>{this.state.dateMessage}</p>}
        <textarea
          placeholder="Add a note for your expense (optional)"
          className="text-area"
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          <button className="button">{this.props.buttonContent ? this.props.buttonContent : 'Add Expense' }</button>
        </div>
      </form>
    );
  }
}
