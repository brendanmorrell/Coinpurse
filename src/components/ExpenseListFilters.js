import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate, sortReverse, setMinAmount, setMaxAmount } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
    dateMessage: null,
  };
  onDatesChange = ({ startDate, endDate }) => { // he says an object gets passed into this that he destructures. try this without destructuring to see if you understand
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
    if (startDate === null && endDate === null) {
      return this.setState(() => ({ dateMessage: 'Date format is DD/MM/YYYY' }));
    }
    return this.setState(() => ({ dateMessage: null }));
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onSortTypeChange = (e) => {
    if (e.target.value === 'date') {
      return this.props.sortByDate();
    }
    return this.props.sortByAmount();
  };
  onSortReverse = () => {
    this.setState((prevState) => ({ filtersReversed: !prevState }));
    this.props.sortReverse();
  };
  onMinChange = (e) => {
    const minAmount = e.target.value;
    if (!minAmount || minAmount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.props.setMinAmount(minAmount);
    }
  };
  onMaxChange = (e) => {
    const maxAmount = e.target.value;
    if (!maxAmount || maxAmount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.props.setMaxAmount(maxAmount);
    }
  };
  setTextFilter = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  clearAllFilters = () => {
    this.props.setTextFilter('');
    this.props.setStartDate(null);
    this.props.setEndDate(null);
    this.props.sortByDate();
    this.props.setMinAmount('');
    this.props.setMaxAmount('');
    if (this.props.filters.sortReverse) {
      return this.props.sortReverse();
    }
    return undefined;
  };
  render() {
    return (
      <div>
        <div className="content-container">
          <div className="search-div">
            <div className="input-group__item">
              <input
                type="text"
                className="text-input"
                placeholder="Search expenses"
                value={this.props.filters.text}
                onChange={this.setTextFilter}
              />
            </div>
          </div>
          <div className="input-group">
            <div className="minmax-div">
              <input
                className="text-input-minmax"
                type="text"
                placeholder="Min"
                value={this.props.filters.minAmount}
                onChange={this.onMinChange}
              />
              <input
                className="text-input-minmax"
                type="text"
                placeholder="Max"
                value={this.props.filters.maxAmount}
                onChange={this.onMaxChange}
              />
              <div className="input-filter-combo">
                <select
                  className="select select-mozilla"
                  value={this.props.filters.sortBy}
                  onChange={this.onSortTypeChange}
                >
                  <option className="option" value="date">Date</option>
                  <option className="option" value="amount">Amount</option>
                </select>
                <button
                  className={this.props.filters.sortReverse ? "filters-reversed input-filter-button" : "filters-normal input-filter-button"}
                  onClick={this.onSortReverse}
                />
              </div>
            </div>
            <div>
              <DateRangePicker
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
              />
            </div>
          </div>
          {this.state.dateMessage && <p className="date-message justify-right ">{this.state.dateMessage}</p>}
          <div className="input-group__button">
            <button
              className="button--clear-filters"
              onClick={this.clearAllFilters}
            >Clear All Filters
            </button>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  sortReverse: () => dispatch(sortReverse()),
  setMinAmount: (minAmount) => dispatch(setMinAmount(minAmount)),
  setMaxAmount: (maxAmount) => dispatch(setMaxAmount(maxAmount)),
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
