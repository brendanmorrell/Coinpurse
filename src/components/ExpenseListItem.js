import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

import { connect } from 'react-redux';

export const ExpenseListItem = ({ id, description, amount, createdAt, note, expenses }) => {
  const formattedAmount = numeral(amount / 100).format('$0, 0.00');

  // Only show year if different from current year
  const displayHours = (expenses.filter((expense) => moment(expense.createdAt).format('D') === moment(createdAt).format('D')).length > 1);
  let dateDisplay;
  const createdThisYear = moment(createdAt).isSame(moment(), 'year');
  if (createdThisYear) {
    if (displayHours) {
      dateDisplay = moment(createdAt).format('MMM Do h:mm a');
    } else {
      dateDisplay = moment(createdAt).format('MMMM Do');
    }
  } else if (!createdThisYear) {
    if (displayHours) {
      dateDisplay = moment(createdAt).format('MMM Do, YYYY h:mm a');
    } else {
      dateDisplay = moment(createdAt).format('MMMM Do, YYYY');
    }
  }

  return (
    <Link className="list-item" to={`/edit/${id}`}>
      <div
        className="list-item-div-one"
      >
        <h3 className="list-item__title">{ description }</h3>
        <h3 className="list-item__subtitle">{ dateDisplay }</h3>
      </div>
      {note &&
        <div className="list-item-note">
          <span className="list-item__subtitle">{note}</span>
        </div>}
      <h3 className="list-item__data">{ formattedAmount }</h3>
    </Link>
  );
};
const mapStateToProps = (state) => ({ expenses: state.expenses });
export default connect(mapStateToProps)(ExpenseListItem);
