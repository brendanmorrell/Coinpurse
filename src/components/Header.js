import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../actions/auth';


export const Header = ({ dispatchStartLogoutProp, currentUser }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <NavLink className="header__title" to="/dashboard" exact={true}>
          <h1>Expensify</h1>
        </NavLink>{/* DOn''t use it here, but NavLink has a bunch of different props that allow you to manipulate the link in cool ways */}
        <div className="header_logout-justify">
          <span className="header__displayName">{currentUser.displayName || currentUser.email}</span><br />
          <button className="button button--link" onClick={dispatchStartLogoutProp}> Logout</button>
        </div>
      </div>
    </div>
  </header>
);
const mapStateToProps = (state) => ({
  currentUser: state.auth.currentUser,
});
const mapDispatchToProps = (dispatch) => ({
  dispatchStartLogoutProp: () => dispatch(startLogout()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
