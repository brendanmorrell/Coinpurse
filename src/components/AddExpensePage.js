import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/dashboard');// this is an included action on react components. you can see it in the react devtools. PUSh takes a single argument which is the address you want to go to, just like links
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);// mapDispatch to props allows us to put dispatch calls as props, thus allowing us to grab them as calls during testing as opposed
