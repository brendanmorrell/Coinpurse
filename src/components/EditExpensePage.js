import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';


export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);// here and below we can use props.match or props.expense since they are both being sent into the EditExpensePage component
    this. props.history.push('/dashboard');
  };
  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/dashboard');
  };
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
          buttonContent="Submit Edit"
        />
        <button
          onClick={this.onRemove}
        >Remove
        </button>
      </div>
    );
  };
};

const mapStateToProps = (state, props) => ({expense: state.expenses.find((expense) => expense.id === props.match.params.id)});

const mapDispatchToProps = (dispatch, props) => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
