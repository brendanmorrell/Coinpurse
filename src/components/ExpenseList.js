import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
// adding 'export' is so you can grap the non-connected component for testing. the default export is the one you actually use for the app, and it is connected to the state via redux



export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>No Expenses currently</p>
      ) : (
        props.expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        })
      )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(ExpenseList);// with connect, you put in the state info you want access to as the first argument, and this returns a function that you then input an argument in using () and insert the component you want to connect to the store.
// the syntax is a bit weird, but the component is placed outside the connect function
