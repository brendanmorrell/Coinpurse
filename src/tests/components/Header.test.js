// had to install enzyme and enzyme-adapter-react-(put your react version number here, like '16') and raf@3.3.2 (request animation frame)
import React from 'react';
// import ReactShallowRenderer from 'react-test-renderer/shallow';// just renders the given component. full rendering would deal with rendering all the related components for testing things like child componenets for user interaction, etc
// ^(this no longer necessary with Enzyme)
import { shallow } from 'enzyme';
// import toJSON from 'enzyme-to-json';
// ^no longer necessary when enzyme-to-jason's snapshot serializer is set in the config file
import { Header } from '../../components/Header';

// redo this now with enzyme
// test('should render Header correctly', () => {
//   const renderer = new ReactShallowRenderer();
//   renderer.render(<Header />);
//   const output = renderer.getRenderOutput();
//   expect(output).toMatchSnapshot(); // this is a jest command that takes a snapshot of the rendered output and saves it to a file (__snapshots__ folder). It compares the current snapshot with whatever the previous rendered snapshot was (thus, the first call to this will always pass, because it just creates a new one which will obs match itself)
// });

let dispatchStartLogoutPropSpy, wrapper;

beforeEach(() => {
  dispatchStartLogoutPropSpy = jest.fn();
  wrapper = shallow(<Header dispatchStartLogoutProp={dispatchStartLogoutPropSpy} />);
});

test('should render Header correctly', () => {
  // const wrapper = shallow(<Header startLogout={() => {}} />);// you pass in the JSX you want to shallow render into this wrapper variable
  // REACTSHALLOW RENDER SETUP (the code below is basic, and you don't need to wrapper line, above. Also not very useful. Enzyme is much better)

  //  expect(wrapper.find('h1').length).toBe(1);
  //  expect(wrapper.find('h1').text()).toBe('Expensify!');

  // ENZYME SETUP EXAMPLE
  // expect(toJSON(wrapper)).toMatchSnapshot();// using this with enzyme now adds a bunch of enzyme scripot into the snapshot, which we don't want because if the enzyme script changes, it will say the snapshot doesn't match, even if what we actually rendered didn't change. Thus, we need to add a library called enzyme-to-json and use it when we are expecting something about our enzyme wrapper
  // ^ with enzyme, need to throw toJSON around the wrapper, and it will extract the meaningful info, and disregard the enzyme content
  // ^you can also wire this up in the jest.config.json filt to run automatically so you don't have to import it or call it
  // ^ you do this by adding:
  // "snapshotSerializers": [
  //   "enzyme-to-json/serializer"
  // ]
  // to the setup file, before or after the "setupFiles" key and then you can run the file as it is written below without importing
  expect(wrapper).toMatchSnapshot();
});


test('expect dispatchStartLogout to have been called on click', () => {
  wrapper.find('button').simulate('click');
  expect(dispatchStartLogoutPropSpy).toHaveBeenCalled();
});
