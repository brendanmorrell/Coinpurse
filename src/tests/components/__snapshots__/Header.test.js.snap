// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <NavLink
        activeClassName="active"
        ariaCurrent="true"
        className="header__title"
        exact={true}
        to="/dashboard"
      >
        <h1>
          Coinpurse
        </h1>
      </NavLink>
      <button
        className="button button--link"
        onClick={[Function]}
      >
         Logout
      </button>
    </div>
  </div>
</header>
`;
